### PR TITLE
[POC] Auto-gen all the python code

### PR DESCRIPTION
**Pull request by logan-stytch**
# Auto-generate all the things

READY FOR REVIEW!
The codegen is run with `bin/generate-api.sh` (which calls into `python -m codegen.generate` with the default arguments). **You need to activate your `virtualenv` first** (unless you get lucky and your current python has necessary dependencies). I think codegen depends on python 3.8, but it may be a newer version… I just used whatever tricks I wanted, but the actual `stytch` library doesn't use any fancy tricks; still need to test for the minimum runnable version, though.

## The spec

You can see what the `.yml` format looks like at `codegen/specs/stytch/`. The sessions file is probably the most “interesting” since it uses a lot of the features — `codegen/specs/stytch/sessions.yml`. The template files are located in `codegen/types/templates/`.

## Client overhaul

The actual client had a massive overhaul to most modules to [1] not use circular imports and [2] separate generated code from “core” code. The client itself had no calling semantics changed — same parameters, same fields so you can do stuff like call `client.otps.sms.send(…)`. In addition, every method now has a second async version which is just `await client.otps.sms.send_async(...)`. To see how the manual sections work, you can check out `stytch/api/sessions.py` and search for `MANUAL`.

## `stytch/core`

The directory `stytch/core/` contains most of what isn’t autogenerated. `api_base.py` has logic to get the full sub-route path for a method, `models.py` include common models (like a `StytchSession`), and `http/client.py` defines a `SyncClient` and `AsyncClient` with associated `get`/`put`/`post`/`delete` methods — it is a thin wrapper around `requests`/`aiohttp` except it bakes the `headers` and `auth` into the `__init__` method.

Also related to `models.py`, every method returns a response that inherits from `ResponseBase` which gives it a `status_code: int` and `request_id: Optional[str]` -> unfortunately `Optional` since `sessions/jwks` doesn’t yet have a `request_id` in its response. It also defines helpful properties like `.is_success` and `.is_client_error` which simply check the `status_code`. Whenever a method is called, we attempt to unmarshal it into its specific type and if that fails, we instead marshal it into a `stytch.core.models.StytchError` type — which also helpfully inherits from `ResponseBase` so you get the same inherited methods.

## Testing

You can run all tests with `python -m unittest` from the root dir, which is very good out-of-the-box and eliminates the need for the `pytest` dependency. To run the integration test, you need to add three env variables: 
- `STYTCH_PYTHON_RUN_INTEGRATION_TESTS=1`
- `STYTCH_PROJECT_ID=your-stytch-project-id`
- `STYTCH_SECRET=your-stytch-secret`
 
There's a helper utility to run all the tests with `bin/generate-coverage.sh` which will also generate a coverage report which is really nifty for viewing actual lines covered or uncovered.

## Is this tested *well*?

Yes, *far* better than the existing library. Overall test coverage is reported as 65% when you omit test files (which misleadingly increase coverage stats). This is in comparison to the existing implementation which has ~44% coverage. Perhaps even better, when looking at _non_ generated code, coverage is 99% (only 6 lines are uncovered). There are details in a comment below

### CODE REVIEW

**Comment by logan-stytch** [comment]
Coverage report:
```
Name                                           Stmts   Miss  Cover
------------------------------------------------------------------
codegen/__init__.py                                0      0   100%
codegen/core/__init__.py                           0      0   100%
codegen/core/snippet_set.py                       67      1    99%
codegen/core/test/__init__.py                      0      0   100%
codegen/core/test/test_snippet_set.py             60      0   100%
codegen/types/__init__.py                          0      0   100%
codegen/types/api.py                              70     26    63%
codegen/types/argument.py                         35      1    97%
codegen/types/http_method.py                      21      1    95%
codegen/types/method.py                           59      4    93%
codegen/types/response_type.py                    20      4    80%
codegen/types/templates/__init__.py                5      2    60%
codegen/types/test/__init__.py                     0      0   100%
codegen/types/test/test_api.py                    45      0   100%
codegen/types/test/test_argument.py               35      0   100%
codegen/types/test/test_http_method.py            12      0   100%
codegen/types/test/test_method.py                 60      0   100%
stytch/__init__.py                                 3      0   100%
stytch/api/__init__.py                             0      0   100%
stytch/api/crypto_wallets.py                      60     29    52%
stytch/api/magic_links.py                         70     41    41%
stytch/api/magic_links_email.py                  137    117    15%
stytch/api/oauth.py                               42     19    55%
stytch/api/otp.py                                 52     22    58%
stytch/api/otp_email.py                           68     35    49%
stytch/api/otp_sms.py                             68     35    49%
stytch/api/otp_whatsapp.py                        68     42    38%
stytch/api/passwords.py                          102     49    52%
stytch/api/passwords_email.py                     84     69    18%
stytch/api/passwords_existing_password.py         38     16    58%
stytch/api/passwords_session.py                   30     17    43%
stytch/api/sessions.py                           107     63    41%
stytch/api/totps.py                               88     43    51%
stytch/api/users.py                              221    135    39%
stytch/api/webauthn.py                            76     57    25%
stytch/client.py                                  36      0   100%
stytch/core/__init__.py                            0      0   100%
stytch/core/api_base.py                           10      0   100%
stytch/core/http/client.py                        46      8    83%
stytch/core/models.py                            119     10    92%
stytch/models/__init__.py                          0      0   100%
stytch/models/crypto_wallets.py                   12      0   100%
stytch/models/magic_links.py                      17      0   100%
stytch/models/magic_links_email.py                13      0   100%
stytch/models/oauth.py                            13      0   100%
stytch/models/otp.py                              10      0   100%
stytch/models/otp_email.py                         8      0   100%
stytch/models/otp_sms.py                           8      0   100%
stytch/models/otp_whatsapp.py                      8      0   100%
stytch/models/passwords.py                        20      0   100%
stytch/models/passwords_email.py                   6      0   100%
stytch/models/passwords_existing_password.py       3      0   100%
stytch/models/passwords_session.py                 5      0   100%
stytch/models/sessions.py                         10      0   100%
stytch/models/totps.py                            26      0   100%
stytch/models/users.py                            50      0   100%
stytch/models/webauthn.py                         18      0   100%
stytch/test/__init__.py                            0      0   100%
stytch/test/test_client.py                        10      0   100%
stytch/version.py                                  1      0   100%
test/__init__.py                                   0      0   100%
test/test_integration.py                         161     25    84%
------------------------------------------------------------------
TOTAL                                           2413    871    64%
```

File: setup.py [hunk 1]
```diff
@@ -33,14 +37,25 @@
...
     ],
-    packages=find_packages(),
+    python_requires=">=3.7",
+    packages=find_packages(
+        include=["stytch*"],
+        exclude=["*.test", "*.tests", "*.test.*", "*.tests.*"],
+    ),
     include_package_data=True,
     install_requires=[
+        "aiohttp>=3.8.3",
```
Comment (jeremy-stytch) [inline setup.py:56]: [pebble] Is there a way to make this dependency optional? It might be annoying for an app to have this extra dependency if they're not using the async API.

File: setup.py [hunk 2]
```diff
@@ -33,14 +37,25 @@
...
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
+        "Programming Language :: Python :: 3.11",
     ],
-    packages=find_packages(),
+    python_requires=">=3.7",
+    packages=find_packages(
+        include=["stytch*"],
+        exclude=["*.test", "*.tests", "*.test.*", "*.tests.*"],
+    ),
```
Comment (jeremy-stytch) [inline setup.py:53]: Just to confirm because I don't understand Python packaging enough: This is what ensures that we only package the actual `stytch` library and not any of our tests or codegen (or codegen tests), right?

File: codegen/types/templates/api.tmpl [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env python3
```
Comment (jeremy-stytch) [inline codegen/types/templates/api.tmpl:1]: [pebble] It feels weird to me to have a shebang line on a non-executable file. Can we remove this from the template so it doesn't generate into the library files?

File: generate-default.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Comment (jeremy-stytch) [inline generate-default.sh:1]: [dust] Gotta support those NixOS users that only have `/bin/sh` and `/usr/bin/env` 😉 
```suggestion
#!/usr/bin/env bash
```

File: codegen/types/argument.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        if self.map_with:
+            value = f"{self.map_with}({self.name})"
+        elif self.map_with_method:
+            value = f"{self.name}.{self.map_with_method}"
+        return f'{dict_name}["{self.name}"] = {value}'
+
+    @property
+    def nullable(self) -> bool:
+        # This could definitely be better
+        return "Optional" in self.arg_type
```
Comment (jeremy-stytch) [inline codegen/types/argument.py:39]: Any ideas? I can see that `r"^Optional\[.*\]$"` would be (pointlessly?) more robust.

You need this to generate the `if field: request['field'] = field` blocks, right? What if we used something smarter than `dict` for `request`? Something that allowed `request.set_if(field)`?

File: stytch/api/sessions.py [hunk 6]
```diff
@@ -1,132 +1,190 @@
...
     def authenticate_jwt(
         self,
         session_jwt: str,
-        *,
         max_token_age_seconds: Optional[int] = None,
         session_custom_claims: Optional[Dict[str, Any]] = None,
-    ):
-        """Parse a JWT and verify the signature, preferring local verification over remote.
+    ) -> AuthenticateResponse:
+        # TODO: Remember to write a test since this is manually generated
```
Comment (jeremy-stytch) [inline stytch/api/sessions.py:120]: Gentle reminder 🙂

**Review by jeremy-stytch — COMMENTED** [review]
I'm excited for well-typed responses and being able to codegen this library! I think this is going in a great direction 😄 

I skimmed codegen to make sure I had at least an overview-level understanding of what was going on.

I read through `stytch/core` and `stytch/api/sessions.py` (because that's the one with fun manual implementations), and they LGTM.

Let me know if you'd like a focused review on any part of this. I'm ready to approve it once you're happy with it 🚀

File: .github/workflows/pull-request.yml [hunk 7]
```diff
@@ -6,25 +6,31 @@ on:
...
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt
 
-    - name: Check types
-      run: mypy stytch tests
+      - name: Check types
+        run: |
+          yes | mypy --install-types
```
[the remark this answers is not in the document]
Comment (logan-stytch) [reply .github/workflows/pull-request.yml:32]: That's just me not reading the docs closely enough to find that option 😅

File: DEVELOPMENT.md [hunk 8]
```diff
@@ -4,13 +4,23 @@ Thanks for contributing to Stytch's Python library! If you run into trouble, fin
 
 ## Setup
 
-1. Install [Python](https://www.python.org/) version 3.4 or greater.
+1. Install [Python](https://www.python.org/) version 3.6 or greater.
```
[the remark this answers is not in the document]
Comment (logan-stytch) [reply DEVELOPMENT.md:7]: Ahh yes, I wanted to actually run the matrix definition in my github workflow to find what version we support.

File: DEVELOPMENT.md [hunk 9]
```diff
@@ -20,4 +30,4 @@ If you have non-trivial changes you'd like us to incorporate, please open an iss
 
 When you're ready for someone to look at your issue or PR, assign `@stytchauth/client-libraries` (GitHub should do this automatically). If we don't acknowledge it within one business day, please escalate it by tagging `@stytchauth/engineering` in a comment or letting us know in [Slack].
 
-[Slack]: https://join.slack.com/t/stytch/shared_invite/zt-nil4wo92-jApJ9Cl32cJbEd9esKkvyg
+[slack]: https://join.slack.com/t/stytch/shared_invite/zt-nil4wo92-jApJ9Cl32cJbEd9esKkvyg
```
[the remark this answers is not in the document]
Comment (logan-stytch) [reply DEVELOPMENT.md:33]: Ha, my autoformatter did this

File: codegen/core/test/test_snippet_set.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            input_lines = f.read()
+        with open(expected_path) as f:
+            contents = f.read()
+            expected: Union[str, Type[Exception]]
+            if contents.strip() == "SnippetKeyMismatchError":
+                expected = SnippetKeyMismatchError
+            elif contents.strip() == "UnterminatedSnippetError":
+                expected = UnterminatedSnippetError
+            else:
+                expected = contents
```
[the remark this answers is not in the document]
Comment (logan-stytch) [reply codegen/core/test/test_snippet_set.py:50]: Ahh, that makes sense, yeah let me do that.

File: setup.py (hunk 2 above)
Comment (logan-stytch) [reply setup.py:53]: Correct, the codegen and tests aren't relevant to put in the PyPI package

File: codegen/types/templates/api.tmpl (hunk 3 above)
Comment (logan-stytch) [reply codegen/types/templates/api.tmpl:1]: Yeah, fair point, just habit to put it on every new python file

File: setup.py (hunk 1 above)
Comment (logan-stytch) [reply setup.py:56]: Oooooh, that could be fun. Let me see if I can make an extra_features like [async] so if you just want a standard client you do "stytch~=6.0" in your requirements.txt, but for async, you do "stytch[async]~=6.0"

I'll play around with that.

File: codegen/types/argument.py (hunk 5 above)
Comment (logan-stytch) [reply codegen/types/argument.py:39]: I'll think about it. I did realize a simple improvement, though is just `self.arg_type.startswith("Optional[")

File: stytch/api/sessions.py (hunk 6 above)
Comment (logan-stytch) [reply stytch/api/sessions.py:120]: Good call.. yeah, I realized I was missing all the tests for the manual methods :( I thought I brought those in already, but alas.

File: DEVELOPMENT.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeremy-stytch) [reply DEVELOPMENT.md:33]: Looks like your autoformater is either too smart or not smart enough 😛 

Which one do you use? I'm curious about options for auto-formatting Markdown because I've never tried that before.

File: codegen/types/argument.py (hunk 5 above)
Comment (jeremy-stytch) [reply codegen/types/argument.py:39]: Oh yeah, that's really all my regex was doing anyway.

I feel like I should mention that I don't think "fixing" this is important. I was mostly curious about the thoughts behind the comment 🙂

File: codegen/core/test/resources/unterminated.expected [hunk 11]
```diff
@@ -0,0 +1 @@
+UnterminatedSnippetError
```
[the remark this answers is not in the document]
Comment (logan-stytch) [reply codegen/core/test/resources/unterminated.expected:1]: (Noting that this is resolved in favor of the other comment)

File: stytch/api/sessions.py (hunk 6 above)
Comment (logan-stytch) [reply stytch/api/sessions.py:118]: Added tests. New coverage:
```
Name                                           Stmts   Miss  Cover
------------------------------------------------------------------
stytch/api/magic_links_email.py                  137    117    15%
stytch/api/passwords_email.py                     84     69    18%
stytch/api/webauthn.py                            76     57    25%
stytch/api/passwords_session.py                   30     17    43%
stytch/api/otp_whatsapp.py                        68     32    53%
stytch/api/magic_links.py                         70     30    57%
stytch/api/users.py                              221     80    64%
stytch/api/otp_email.py                           68     18    74%
stytch/api/otp_sms.py                             68     18    74%
stytch/api/oauth.py                               42     10    76%
stytch/api/sessions.py                           110     26    76%
stytch/api/crypto_wallets.py                      60     14    77%
stytch/api/otp.py                                 52     12    77%
stytch/api/passwords.py                          102     22    78%
stytch/api/passwords_existing_password.py         38      8    79%
stytch/api/totps.py                               88     18    80%
codegen/types/api.py                              72      6    92%
codegen/__init__.py                                0      0   100%
codegen/core/__init__.py                           0      0   100%
codegen/core/snippet_set.py                       67      0   100%
codegen/types/__init__.py                          0      0   100%
codegen/types/argument.py                         35      0   100%
codegen/types/http_method.py                      21      0   100%
codegen/types/method.py                           59      0   100%
codegen/types/response_type.py                    20      0   100%
codegen/types/templates/__init__.py                5      0   100%
stytch/__init__.py                                 3      0   100%
stytch/api/__init__.py                             0      0   100%
stytch/client.py                                  36      0   100%
stytch/core/__init__.py                            0      0   100%
stytch/core/api_base.py                           10      0   100%
stytch/core/http/client.py                        37      0   100%
stytch/core/models.py                            119      0   100%
stytch/models/__init__.py                          0      0   100%
stytch/models/crypto_wallets.py                   12      0   100%
stytch/models/magic_links.py                      17      0   100%
stytch/models/magic_links_email.py                13      0   100%
stytch/models/oauth.py                            13      0   100%
stytch/models/otp.py                              10      0   100%
stytch/models/otp_email.py                         8      0   100%
stytch/models/otp_sms.py                           8      0   100%
stytch/models/otp_whatsapp.py                      8      0   100%
stytch/models/passwords.py                        20      0   100%
stytch/models/passwords_email.py                   6      0   100%
stytch/models/passwords_existing_password.py       3      0   100%
stytch/models/passwords_session.py                 5      0   100%
stytch/models/sessions.py                         10      0   100%
stytch/models/totps.py                            26      0   100%
stytch/models/users.py                            50      0   100%
stytch/models/webauthn.py                         18      0   100%
stytch/version.py                                  1      0   100%
------------------------------------------------------------------
TOTAL                                           2026    554    73%
```

File: setup.py [hunk 12]
```diff
@@ -56,6 +56,6 @@
         "aiohttp>=3.8.3",
         "requests>=2.7.0",
         "pydantic>=1.10.2",
-        "pyjwt[crypto]>=2.3.0",
+        "pyjwt[crypto]>=2.4.0",
```
Comment (jeremy-stytch) [inline setup.py:59]: Oh, that reminds me: we'll probably be able to close #74 and #83 after this merges

**Review by jeremy-stytch — APPROVED** [review]
LGTM!

I think my only leftover comment is about making the `aiohttp` dependency optional for someone who is fine with sync-only and doesn't want the extra dependency. But I don't see a reason to block all the extra benefits of this new version on that. We can make the dependency optional in a v6.1 if the upgrade path is as easy as consumers changing their deps from `stytch` to `stytch[async]`.

If it's easier to fix the merge conflicts with a (squash-)rebase, I'm totally cool with that. GitHub's code review gets a bit weird after a rebase, but a full-rewrite PR was bound to get messy 😄

File: setup.py (hunk 12 above)
Comment (logan-stytch) [reply setup.py:59]: Yeah, I should've mentioned, I looked into the optional features and couldn't figure out an easy way to divorce the `aiohttp` library completely. For now, I think this is the best I can do. There's probably some way to conditionally exclude the `aiohttp` library, but I don't know what patterns there are typical.
